ACG/AutoRouterShield.py

```diff
@@ -0,0 +1,714 @@
+        relative_coords : bool
+            True if the list of coordinates are relative to the starting port's coordinate.
+            False if the list of coordinates are absolute relative to the current Template's origin
+
+        Returns
+        -------
+        self : WgRouter
+            returns itself so that route segments can be easily chained together
+        """
+        if not self.current_rect or not self.current_handle or not self.current_dir:
+            raise ValueError('Router has not been initialized, please call new_route()')
+
+        points = self.route_points
+
+        current_dir = self.current_dir
+        current_point = (self.current_rect[self.current_handle].xy, self.current_rect.layer)
+
+        if relative_coords:
+            # If passed coordinates are relative, need to add WgRouter's port location to convert to absolute coords
+            x0, y0 = current_point[0]
+            points = [((pt[0] + x0, pt[1] + y0), layer) for pt, layer in points]
+
+        # Generate a manhattanized list of waypoints on the route while minimizing the number of required bends
+        manh_point_list = self.manhattanize_point_list(initial_direction=current_dir,
+                                                       initial_point=current_point,
+                                                       points=points)
+
+        for i in range(len(manh_point_list)):
+            point = manh_point_list[i]
+            if tuple(point[0]) not in self.route_point_dict:
+                if i != 0:
+                    self.route_point_dict[tuple(point[0])] = self.route_point_dict[tuple(manh_point_list[i - 1][0])]
+
+        final_point_list = self.process_manh(manh_point_list)
+
+        # Simplify the point list so that each point corresponds with a bend of the route, i.e. no co-linear points
+        final_point_list = final_point_list[1:]  # Ignore the first pt, since it is co-incident with the starting port
+
+        # Draw a series of L-routes to follow the final simplified point list
+        for pt0, pt1 in zip(final_point_list, final_point_list[1:]):
+            # print(f'drawing route {pt0[0]} -> {pt1[0]} on layer {pt0[1]}')
+            self._draw_route_segment(pt0=pt0,
+                                     pt1=pt1,
+                                     in_width=self.route_point_dict[tuple(pt0[0])],
+                                     out_width=self.route_point_dict[tuple(pt1[0])],
+                                     enc_style='asymm')
+
+        # The loop does not draw the final straight segment, so add it here
+        self._draw_route_segment(pt0=final_point_list[-1],
+                                 pt1=None,
+                                 in_width=self.route_point_dict[tuple(final_point_list[-1][0])],
+                                 out_width=self.route_point_dict[final_point_list[-1][0]],
+                                 enc_style='uniform')
+
+    @staticmethod
+    def overlap(A, B):
```

Comment:
Rectangles already have a built in overlap method, which returns a new rectangle which is the size of the overlapping region. If you want to have a method which just determines whether or not the two rectangles overlap in a certain dimension, we should move this to the Rectangle class